File: Functional.FUnitTests/MatchingTests.fs
module MatchingTests

open NUnit.Framework
open FsUnit
open System
open Psns.Common.Functional

type ex = Psns.Common.Functional.Prelude

let trueMatch value = ex.AsEqual(value, Func<bool, bool> (fun b -> b))
let catcher = Func<bool, Maybe<bool>> (fun b -> ex.Some(b))
let matchFun cond value = ex.Match(cond, trueMatch value, catcher)

[<Test>]
let ``it should match a bool correctly.`` () =
    matchFun (1 > 0) true |> should equal true

File: Functional.UnitTests/MatchingTests.cs
using NUnit.Framework;
using static NUnit.StaticExpect.Expectations;
using static Psns.Common.Functional.Prelude;

namespace SystemExtensions.UnitTests.Functional
{
    [TestFixture]
    public class MatchingTests
    {
        [Test]
        public void Should_Match_Bool() =>
            Expect(
                Match(
                    false, 
                    AsEqual(true, _ => true), 
                    _ => _),
                False);

        [Test]
        public void Should_MatchFirst_Bool() =>
            Expect(
                Match(
                    true,
                    AsEqual(true, _ => true)),
                True);

        [Test]
        public void NoMatch_Should_Throw() =>
            Expect(() =>
                Match(
                    false,
                    AsEqual(true, _ => true)),
                Throws.InvalidOperationException);
    }
}